### PR TITLE
support match case/regex filter in log view

### PR DESCRIPTION
**Pull request by quangson91**
Ref: https://github.com/flutter/flutter-intellij/issues/2200

+ Implement support filter by regex & match case in the log view.
+ Refactoring by creating a form handle filter actions.

Notes:
If user enable `regex` => we will automatically added `.*` before/after the filter text.

Example:
User input: `hello`
=> The actual text filter in mode regex is `.*hello.*`

Screenshot:

@pq the UI will look like: 
flutter_log_filter

### CODE REVIEW

File: src/io/flutter/logging/FlutterLogFilterPanel.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+import javax.swing.*;
+
+
+public class FlutterLogFilterPanel {
+  private JPanel root;
+  private JCheckBox chbMatchCase;
+  private JCheckBox chbRegex;
+  private JPanel pnlExpression;
+  private final LogFilterTextField txtExpression;
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogFilterPanel.java]: nit: could we expand these field names to be less abbreviated?  maybe:

`matchCaseCheckBox`
`regexCheckBox`
`panelExpression` and
`textExpression`

?

File: src/io/flutter/logging/FlutterLogTree.java [hunk 2]
```diff
@@ -266,12 +266,22 @@ protected void customizeCellRenderer(JTable table, @Nullable Object value, boole
 
 
   public static class ContainsTextFilter implements EntryFilter {
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogTree.java]: Since this is more than just matching containing text, perhaps rename it to something more general.  Maybe `LogEntryFilter`?  Or perhaps do away with the interface altogether and just have an `EntryFilter` class?  (I was thinking there'd be a separate implementation for each matching strategy but what you've go works great too. 👍

File: src/io/flutter/logging/FlutterLogTree.java [hunk 3]
```diff
@@ -281,27 +291,36 @@ public boolean accept(@NotNull FlutterLogEntry entry) {
       if (text == null) {
         return true;
       }
-
-      if (entry.getMessage().contains(text)) {
+      final String standartText = isMatchCase ? text : text.toLowerCase();
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogTree.java]: standartText => standardText?

File: src/io/flutter/logging/FlutterLogTree.java [hunk 4]
```diff
@@ -281,27 +291,36 @@ public boolean accept(@NotNull FlutterLogEntry entry) {
...
+      final String standartMessage = isMatchCase ? entry.getMessage() : entry.getMessage().toLowerCase();
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogTree.java]: standart => standard

File: src/io/flutter/logging/FlutterLogTree.java [hunk 5]
```diff
@@ -281,27 +291,36 @@ public boolean accept(@NotNull FlutterLogEntry entry) {
...
+      final String standartCategory = isMatchCase ? entry.getCategory() : entry.getCategory().toLowerCase();
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogTree.java]: standart => standard

File: src/io/flutter/logging/FlutterLogTree.java [hunk 6]
```diff
@@ -281,27 +291,36 @@ public boolean accept(@NotNull FlutterLogEntry entry) {
...
         return true;
       }
+      return acceptByCheckingRegexOption(standartCategory, standartText);
+    }
 
-      //noinspection RedundantIfStatement
-      if (entry.getCategory().contains(text)) {
-        return true;
+    private boolean acceptByCheckingRegexOption(@NotNull String message, @NotNull String text) {
+      final String standardString = message.replaceAll("\n", "");
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogTree.java]: I'm curious, why is this necessary?

File: src/io/flutter/logging/FlutterLogTree.java [hunk 7]
```diff
@@ -281,27 +291,36 @@ public boolean accept(@NotNull FlutterLogEntry entry) {
...
+      final ContainsTextFilter filter = (ContainsTextFilter)o;
+      return isRegex == filter.isRegex &&
+             isMatchCase == filter.isMatchCase &&
+             Objects.equals(text, filter.text);
     }
 
     @Override
     public int hashCode() {
-      return 13 ^ Objects.hashCode(text);
+      return Objects.hash(text, isRegex, isMatchCase);
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogTree.java]: 👍

File: src/io/flutter/logging/FlutterLogView.java [hunk 8]
```diff
@@ -29,77 +31,19 @@
...
-import java.awt.event.KeyEvent;
 import java.awt.event.MouseEvent;
 
 public class FlutterLogView extends JPanel implements ConsoleView, DataProvider, FlutterLog.Listener {
-
   // based on com.intellij.openapi.vcs.ui.SearchFieldAction
   private class FilterToolbarAction extends AnAction implements CustomComponentAction /*, RightAlignedToolbarAction */ {
-    private final JPanel myComponent;
-    private final SearchTextField myField;
+    private final FlutterLogFilterPanel pnlFlutterLogFilter = new FlutterLogFilterPanel(param -> doFilter());
```
Comment (pq) [inline src/io/flutter/logging/FlutterLogView.java]: pnlFlutterLogFilter => filterPanel?

File: src/io/flutter/logging/LogFilterTextField.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  private void doFilterIfNeeded() {
+    if (onFilterListener != null) {
+      onFilterListener.onFilter();
+    }
+  }
+
+  public interface OnFilterListener {
+    void onFilter();
+  }
+}
```
Comment (pq) [inline src/io/flutter/logging/LogFilterTextField.java]: Add a trailing newline?

File: testSrc/unit/io/flutter/logging/FlutterLogTreeTest.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    final FlutterLogTree.ContainsTextFilter filterMatchCaseNormal = new FlutterLogTree.ContainsTextFilter("hello", true, false);
+    final FlutterLogTree.ContainsTextFilter filterMatchCaseRegex = new FlutterLogTree.ContainsTextFilter("hello", true, true);
+    final FlutterLogTree.ContainsTextFilter filterInvalidMatchCaseNormal = new FlutterLogTree.ContainsTextFilter("Hello", true, false);
+    final FlutterLogTree.ContainsTextFilter filterInvalidMatchCaseRegex = new FlutterLogTree.ContainsTextFilter("Hello", true, true);
+    assertTrue(filterMatchCaseNormal.accept(entry));
+    assertTrue(filterMatchCaseRegex.accept(entry));
+    assertFalse(filterInvalidMatchCaseNormal.accept(entry));
+    assertFalse(filterInvalidMatchCaseRegex.accept(entry));
+  }
+}
```
Comment (pq) [inline testSrc/unit/io/flutter/logging/FlutterLogTreeTest.java]: Add a trailing newline?

File: src/io/flutter/logging/FlutterLogTree.java (hunk 2 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: (This is not a blocking ask.  Feel free to ignore and we can do something on another pass.)

**Comment by pq** [comment]
Super.  Thanks for the screenshot @quangson91.  This is great!

I wonder if we can remove the "Expression" label to get a bit more horizontal space?  I think it's pretty clear what the search box is for.  Perhaps we could add a nice tooltip if you think it's unclear?

**Comment by quangson91** [comment]
@pq I take a screenshot from android logcat.
We may discuss more about UI specs for flutter view before we work.
flutter_log_view_specs

**Comment by quangson91** [comment]
>I wonder if we can remove the "Expression" label to get a bit more horizontal space? I think it's pretty >clear what the search box is for. Perhaps we could add a nice tooltip if you think it's unclear?

Yeap. @pq I will do that.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 2 above)
Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: I choose a solution to merge them all in class `EntryFilter`.
If in the future we need more tyle filter -> we will refactor later.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 3 above)
Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: Oops. My typos mistake. I updated.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 4 above)
Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: updated.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 5 above)
Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: Updated.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 6 above)
Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: Because java regex doesn't recognize as a character.
For example:

```java
public class Main {
    public static void main(String args[]) {
        String input = "\n hello world \n";
        String regex = ".*hello.*";
        System.out.println("Result: " + input.matches(regex));
    }
}
```
Result: `false`.

But if we change input to:
```
String input = "\n hello world \n".replaceAll("\n", "");
```
=> Result: `true`.

Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: The affect of this code is user can't filter with string context break line.

File: src/io/flutter/logging/FlutterLogView.java (hunk 8 above)
Comment (quangson91) [reply src/io/flutter/logging/FlutterLogView.java]: Updated. Thanks!

File: src/io/flutter/logging/LogFilterTextField.java (hunk 9 above)
Comment (quangson91) [reply src/io/flutter/logging/LogFilterTextField.java]: Done.

File: testSrc/unit/io/flutter/logging/FlutterLogTreeTest.java (hunk 10 above)
Comment (quangson91) [reply testSrc/unit/io/flutter/logging/FlutterLogTreeTest.java]: Done.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 6 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: Aha.  Thanks!  If this is just for the regex case, should we move the `replaceAll` into the if block?

IOW, something like:
```java
if (isRegex) {
  // Extra replace to support matching on strings w/ line breaks.
  return message.replaceAll("\n", "").matches(".*" + text + ".*");
}
```

Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: Yeap. I updated. Thank you!

Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: Cool!  One last question...  I'm rusty w/ regexps but I wonder if we there isn't some combination of Pattern flags we couldn't use to get around need to modify the message string.  Maybe look at [`DOTALL`](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#DOTALL) and [`MULTILINE`](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#MULTILINE)?

Happy to land an iterate as well.

Thanks again!

Comment (quangson91) [reply src/io/flutter/logging/FlutterLogTree.java]: Yes, It looks better.
We just need to add `(?s)` to support DOTALL.
```
return message.matches("(?s).*" + text + ".*");
```
That's great!.
Thank you!

**Comment by quangson91** [comment]
@pq , There is something coming up for logging feature.

Ref: https://i.stack.imgur.com/q3dEU.gif

+ Log Level
+ Color by log level (can custom color by regex too).
+ Filter (match case, regex)
+ Integrate with GrepConsole (grep console only support log view extends from com.intellij.execution.impl.ConsoleViewImpl).
+ Support multi tab for filter. 
+ Clickable in reference in console. (Eg. In console we have xyz.dart:15 -> click open xyz.dart file)
+ Auto scroll at the end / Keep scroll position.
+ Append line able. (Eg. we have log 1, 2, 3, 4. user click append line => we append "=========" then the comming log will bellow "======";)
+ Copyable.
+ Wrap line option.
+ Folding option.

===
Other
+ select log by app.
+ select log by device.

**Comment by pq** [comment]
@quangson91 : great list.  We do have a bunch of these already and a few are already in progress, but others aren't planned yet.  Maybe call out things you'd like to see happen on #2200 so we can discuss and coordinate after this PR is merged? I'd also like to be careful not to turn the logging view into a feature kitchen sink.  In particular, I think the prior art of loggers isn't the best model for usability and would like to start from the use cases rather from the implementations.  Anyway, yeah.  Exciting stuff.  Thanks for your continued contributions!

**Comment by quangson91** [comment]
@pq, yeap, I got it.
